feat(search): add youtube search

OCannings opened the pull request:
A first attempt at integrating the Youtube search API has been made.
This is a work in progress and currently untested.

An environment variable has been introduced: YOUTUBE_API_KEY

Visit http://localhost:4000/api/videos?q=hello to see a list of videos
pulled from Youtube.

TODO: write automated tests and handle API failures/malformed JSON
responses

OCannings committed:
feat(search): add youtube search

A first attempt at integrating the Youtube search API has been made.
This is a work in progress and currently untested.

An environment variable has been introduced: YOUTUBE_API_KEY

Visit http://localhost:4000/api/videos?q=hello to see a list of videos
pulled from Youtube.

TODO: write automated tests and handle API failures/malformed JSON
responses

ljkbennett commented on config/dev.exs:
```diff
@@ -41,3 +41,6 @@ config :juicebox, Juicebox.Repo,
   pool_size: 10
+
+config :juicebox, Juicebox.Youtube,
+  api_key: ""
```
why is the dev version empty?

OCannings replied:
I just removed my api key before committing. Need something better here.

OCannings committed:
reafactor(search): replace HTTPotion -> HTTPoison

ljkbennett commented:
I set up an API key and added it to dev.exs directly.
Getting an error message: `protocol Enumerable not implemented for nil` on line 7 of api_controller.ex

OCannings committed:
feat(search): add basic error handling

ANY error will trigger a "Unable to fetch results" response.

OCannings committed:
fix(search): handle errors from Youtube API

e.g. Invalid API key

OCannings commented:
@littleowllabs Looks like you have an invalid API key. I just updated the pr to handle that case.

ljkbennett commented:
Just worked it out. I had a valid API key, but I hadn't enabled the API 😿

OCannings committed:
refactor(youtube): move youtube client to config

The youtube api client can now be set in configuration, this will allow
us to use mock versions in test.

Two implementations have been added, one "real" http client and another
that returns dummy data for tests.

See: http://blog.plataformatec.com.br/2015/10/mocks-and-explicit-contracts/

OCannings committed:
refactor(youtube): add Video struct

OCannings committed:
test(api): add basic Api controller tests

ljkbennett commented on config/dev.exs:
```diff
@@ -41,3 +41,8 @@ config :juicebox, Juicebox.Repo,
+config :juicebox, :youtube_api, Juicebox.Youtube.HTTP
+
+config :juicebox, Juicebox.Youtube,
+  api_key: ""
```
why not put the api key into an environment variable for dev too?

OCannings committed:
chore(config): add YT api key as env var in dev

ljkbennett commented on lib/juicebox/youtube/http.ex:
```diff
@@ -0,0 +1,73 @@
+          }
+        }
+      }
+    } = video
```
Is this the preferred approach rather than doing it in the definition when there is a big pattern to match?

OCannings replied:
Like this?

``` elixir
defp format_video(%{
      "id" => %{
        "videoId" => video_id
      },
      "snippet" => %{
        "title" => title,
        "description" => description,
        "thumbnails" => %{
          "high" => %{
            "url" => thumbnail
          }
        }
      }
    }) do

    %Video{
       video_id: video_id,
       title: title,
       description: description,
       thumbnail: thumbnail
     }
  end
```

ljkbennett replied:
Yeah, was just thinking aloud really. I think it's ugly, but it'd not much less ugly than the other way and saves extra intermediate allocations.
